core/src/main/java/org/apache/iceberg/util/JsonUtil.java

```diff
@@ -116,19 +122,42 @@ public static String getStringOrNull(String property, JsonNode node) {
   }
 
   public static List<String> getStringList(String property, JsonNode node) {
-    Preconditions.checkArgument(node.has(property), "Cannot parse missing list %s", property);
+    ImmutableList.Builder<String> builder = ImmutableList.builder();
+    return getCollection(property, node, builder::build, builder::add, JsonNode::asText,
+        e -> Preconditions.checkArgument(e.isTextual(), "Cannot parse string from non-text value: %s", e),
+        false);
+  }
+
+  public static Set<Integer> getIntegerSetOrNull(String property, JsonNode node) {
+    ImmutableSet.Builder<Integer> builder = ImmutableSet.builder();
+    return getCollection(property, node, builder::build, builder::add, JsonNode::asInt,
+        e -> Preconditions.checkArgument(e.isInt(), "Cannot parse string from non-integer value: %s", e),
+        true);
+  }
+
+  private static <C extends Collection<T>, T> C getCollection(
+      String property, JsonNode node,
+      Supplier<C> builder, Consumer<T> appender,
+      Function<JsonNode, T> getter, Consumer<JsonNode> validator,
```

Comment:
Those `builder`, `appender`, `getter` and `validator` makes the method to be really hard to follow, could we just defines a simple `visitor` interface to visit the `JsonNode` in this collection and collect all the final result to the upper layer method ?

Reply:
thanks for the suggestion, please see if the new approach looks better.